source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,342 @@
+
+  switch (state_) {
+
+  case State::INITIAL: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send success callback when no request is in progress");
+    break;
+  }
+  case State::CREATE_CONFIG: {
+    // get the config object that was created
+    if (m->headers().Status()->value() != createdCode().c_str()) {
+      ENVOY_LOG(info, "Squash: can't create attachment object. status {} - not squashing",
+                m->headers().Status()->value().c_str());
+      doneSquashing();
+    } else {
+      state_ = State::CHECK_ATTACHMENT;
+
+      std::string debugAttachmentId;
+      try {
+        Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+        debugAttachmentId =
+            json_config->getObject("metadata", true)->getString("name", EMPTY_STRING);
+      } catch (Json::Exception&) {
+        debugAttachmentId = EMPTY_STRING;
+      }
+
+      if (debugAttachmentId.empty()) {
+        ENVOY_LOG(info, "Squash: failed to parse debug attachment object - check server settings.");
+        doneSquashing();
+      } else {
+        debugAttachmentPath_ = postAttachmentPath() + debugAttachmentId;
+        pollForAttachment();
+      }
+    }
+
+    break;
+  }
+  case State::CHECK_ATTACHMENT: {
+
+    std::string attachmentstate;
+    try {
+      Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+      attachmentstate = json_config->getObject("status", true)->getString("state", EMPTY_STRING);
+    } catch (Json::Exception&) {
+      // no state yet.. leave it empty for the retry logic.
+    }
+
+    bool attached = attachmentstate == attachedState();
+    bool error = attachmentstate == errorState();
+    bool finalstate = attached || error;
+
+    if (finalstate) {
+      doneSquashing();
+    } else {
+      retry();
+    }
+    break;
+  }
+  }
+}
+
+void SquashFilter::onFailure(AsyncClient::FailureReason) {
+  // in_flight_request_ will be null if we are called inline of async client send()
+  bool request_created = in_flight_request_ != nullptr;
+  in_flight_request_ = nullptr;
+  switch (state_) {
+  case State::INITIAL: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send failure callback when no request is in progress");
+    break;
+  }
+  case State::CREATE_CONFIG: {
+    // no retries here, as we couldnt create the attachment object.
+    if (request_created) {
+      // cleanup not needed if onFailure called inline in async client send.
+      // this means that decodeHeaders is down the stack and will return Continue.
+      doneSquashing();
+    }
+    break;
+  }
+  case State::CHECK_ATTACHMENT: {
+    retry();
+    break;
+  }
+  }
+}
+
+void SquashFilter::retry() {
+
+  if (delay_timer_.get() == nullptr) {
+    delay_timer_ =
+        decoder_callbacks_->dispatcher().createTimer([this]() -> void { pollForAttachment(); });
+  }
+  delay_timer_->enableTimer(config_->attachment_poll_period());
+}
+
+void SquashFilter::pollForAttachment() {
```

Comment:
Drive-by-comment: have you considered using gRPC streaming for this protocol instead of JSON polling? I realize this is beyond the scope of the review, but it seems like it would be architecturally cleaner.

Reply:
I agree, and that is a good idea for a squash enhancement in a future release.